Add support for `inline` option when defining `@theme` values

adamwathan opened the pull request:
This PR adds support for a new `inline` option when defining a `@theme` block that tells Tailwind to use raw theme values for utilities instead of referencing the corresponding generated CSS variable.

```css
/* Input */
@theme inline {
  --color-red-500: #ef4444;
  /* ... */
}


/* Example output */
:root {
  --color-red-500: #ef4444;
}

.text-red-500 {
  color: #ef4444;
}
```

This can be composed with the existing `reference` option in case you want to define a `@theme` block as both `reference` (so the variables aren't generated) and `inline`:

```css
/* Input */
@theme inline reference {
  --color-red-500: #ef4444;
  /* ... */
}


/* Example output */
.text-red-500 {
  color: #ef4444;
}
```

Since you can have multiple `@theme` blocks, you can even define some values normally and some as inline based on how you're using them. For example you might want to use `inline` for defining literal tokens like `--color-red-500`, but include the variable for tokens that you want to be able to theme like `--color-primary`:

```css
/* Input */
@theme inline {
  --color-red-500: #ef4444;
  /* ... */
}

@theme {
  --color-primary: var(--color-red-500);
}


/* Example output */
:root {
  --color-red-500: #ef4444;
  --color-primary: var(--color-red-500);
}

.text-red-500 {
  color: #ef4444;
}

.text-primary {
  color: var(--color-primary, var(--color-red-500));
}
```

## Breaking changes

Prior to this PR, you could `@import` a stylesheet that contained `@theme` blocks as reference by adding the `reference` keyword to your import:

```css
@import "./my-theme.css" reference;
```

Now that `reference` isn't the only possible option when declaring your `@theme`, this syntax has changed to a new `theme(…)` function that accepts `reference` and `inline` as potential space-separated values:

```css
@import "./my-theme.css";
@import "./my-theme.css" theme(reference);
@import "./my-theme.css" theme(inline);
@import "./my-theme.css" theme(reference inline);
```

If you are using the `@import … reference` option with an earlier alpha release, you'll need to update your code to `@import … theme(reference)` once this PR lands in a release.

## Motivation

This PR is designed to solve an issue pointed out in #14091.

Prior to this PR, generated utilities would always reference variables directly, with the raw value as a fallback:

```css
/* Input */
@theme {
  --color-red-500: #ef4444;
  /* ... */
}


/* Example output */
:root {
  --color-red-500: #ef4444;
}

.text-red-500 {
  color: var(--color-red-500, #ef4444);
}
```

But this can create issues with variables resolving to an unexpected value when a theme value is referencing another variable defined on `:root`.

For example, say you have a CSS file like this:

```css
:root, .light {
  --text-fg: #000; 
}

.dark {
  --text-fg: #fff;
}

@theme {
  --color-fg: var(--text-fg);
}
```

Without `@theme inline`, we'd generate this output if you used the `text-fg` utility:

```css
:root, .light {
  --text-fg: #000; 
}

.dark {
  --text-fg: #fff;
}

:root {
  --color-fg: var(--text-fg);
}

.text-fg {
  color: var(--color-fg, var(--text-fg));
}
```

Now if you wrote this HTML, you're probably expecting your text to be the dark mode color:

```html
<div class="dark">
  <h1 class="text-fg">Hello world</h1>
</div>
```

But you'd actually get the light mode color because of this rule:

```css
:root {
  --color-fg: var(--text-fg);
}

.text-fg {
  color: var(--color-fg, var(--text-fg));
}
```

The browser will try to resolve the `--color-fg` variable, which is defined on `:root`. When it tries to resolve the value, _it uses the value of `var(--text-fg)` as it would resolve at `:root`_, not what it would resolve to based on the element that has the `text-fg` class.

So `var(--color-fg)` resolves to `#000` because `var(--text-fg)` resolved to `#000` at the point in the tree where the browser resolved the value of `var(--color-fg)`.

By using `@theme inline`, the `.text-fg` class looks like this:

```css
.text-fg {
  color: var(--text-fg);
}
```

With this definition, the browser doesn't try to resolve `--color-fg` at all and instead resolves `--text-fg` directly which correctly resolves to `#fff` as expected.

adamwathan committed:
Add support for `inline` option when defining `@theme` values

RobinMalfait commented on packages/tailwindcss/src/index.test.ts:
```diff
@@ -1120,9 +1120,135 @@ describe('Parsing themes values from CSS', () => {
+        ['bg-tomato', 'bg-potato', 'bg-primary'],
+      ),
+    ).toMatchInlineSnapshot(`
+      ".bg-potato {
```
I think this should contain the `:root` with the variables. It currently behaves as if we wrote `@media theme(inline reference)`.

```css
:root {
  --color-tomato: #e10c04;
  --color-potato: #ac855b;
  --color-primary: var(--primary);
}
```

adamwathan replied:
Good catch, fixed! (By you, really 😄)

adamwathan committed:
Skip walking children of `@media theme(…)`

RobinMalfait commented on packages/tailwindcss/src/index.ts:
```diff
@@ -152,28 +152,35 @@ export function compile(
-    if (node.selector !== '@theme' && node.selector !== '@theme reference') return
+    if (node.selector !== '@theme' && !node.selector.startsWith('@theme ')) return
+
+    let [_, ...themeOptions] = segment(node.selector, ' ')
```
We are not doing anything with the `themeOptions` as a whole except for checking if the `inline` or `reference` exists. So we can just keep the `@media` in there which saves us some destructuring stuff.
```suggestion
    let themeOptions = segment(node.selector, ' ')
```

RobinMalfait commented on packages/tailwindcss/src/index.ts:
```diff
@@ -395,15 +402,17 @@ export function __unstable__loadDesignSystem(css: string) {
-    if (node.selector !== '@theme' && node.selector !== '@theme reference') return
-    let isReference = node.selector === '@theme reference'
+    if (node.selector !== '@theme' && !node.selector.startsWith('@theme ')) return
+    let [_, ...themeOptions] = segment(node.selector, ' ')
```
```suggestion
    let themeOptions = segment(node.selector, ' ')
```

RobinMalfait commented on packages/tailwindcss/src/index.ts:
```diff
@@ -152,28 +152,35 @@ export function compile(
+    let [_, ...themeOptions] = segment(node.selector, ' ')
 
-    let isReference = node.selector === '@theme reference'
+    let isReference = themeOptions.includes('reference')
+    let isInline = themeOptions.includes('inline')
```
If you apply the suggestion from above and if you are worried about the `includes` that it needs to check `@media` (first argument of the array), then you can skip it by doing something like this:

```suggestion
    let isReference = themeOptions.includes('reference', 1)
    let isInline = themeOptions.includes('inline', 1)
```
Ultimate performance!

thecrypticace replied:
TIL I had no idea includes took a starting position

RobinMalfait commented on packages/tailwindcss/src/index.ts:
```diff
@@ -395,15 +402,17 @@ export function __unstable__loadDesignSystem(css: string) {
-    let isReference = node.selector === '@theme reference'
+    if (node.selector !== '@theme' && !node.selector.startsWith('@theme ')) return
+    let [_, ...themeOptions] = segment(node.selector, ' ')
+    let isReference = themeOptions.includes('reference')
+    let isInline = themeOptions.includes('inline')
```
```suggestion
    let isReference = themeOptions.includes('reference', 1)
    let isInline = themeOptions.includes('inline', 1)
```

RobinMalfait committed:
improve themeOptions usage

Destructuring works, but it's creating an unnecessary array. We are
currently interested in whether the `inline` and/or `reference` values
exist in the `themeOptions` list.

We already know that the very first item is going to be `@theme`, so we
can skip over that by providing the `fromIndex` when using
`.includes(value, fromIndex)`

RobinMalfait committed:
update changelog

adamwathan committed:
Extract parseThemeOptions function

adamwathan commented on packages/tailwindcss/src/index.ts:
```diff
@@ -33,6 +33,21 @@ function throwOnPlugin(): never {
   throw new Error('No `loadPlugin` function provided to `compile`')
 }
 
+function parseThemeOptions(selector: string) {
+  let isReference = false
+  let isInline = false
+
+  for (let option of segment(selector.slice(6) /* '@theme'.length */, ' ')) {
+    if (option === 'reference') {
+      isReference = true
+    } else if (option === 'inline') {
+      isInline = true
+    }
+  }
+
+  return { isReference, isInline }
+}
+
```
@RobinMalfait Decided to extract this function because the work was being done twice, and `themeOptions` didn't feel like a good name for an array that contained:

```
['@theme', 'reference', 'inline']
```

Refactored to a `for` loop to avoid the extra loop that calling `includes` twice would incur, but admittedly sad it's more code.
```

RobinMalfait replied:
Nice, makes sense.

One potential thing we can do is if both options are `true` early return because there is no need to keep looping if everything is found already. But since `@theme` is so specific to Tailwind, there is no need for that optimization (yet).

marklawlor commented:
It is possible to apply `@theme inline reference` to the default theme or is only available for theme overrides?

The use case is [NativeWind](https://nativewind.dev) which is a JS implementation of CSS for React Native. While it can handle the CSS variables used in the theme, it would be more performant to inline them. 

The workaround currently is to either use `postcss`/`lightningcss` to inline the variables or override every theme variable with `@theme inline reference`

abhchand commented:
Keep in mind that **prefixes apply to variable names referenced in `@theme inline`**.

(For anyone anyone using inline themes and prefixes, but seeing `--(var name) is not set` in the browser, you may not be referencing the right css var)

Example, suppose you use the namespace `tw` in tailwind v4:

```js
// tailwind.config.js
module.exports = {
  ...
  prefix: 'tw',
  ...
};
```

You'd set `inline` definitions as follows:

```css
// index.css

@theme {
  --color-flickr-blue: #0066DB;
}

@theme inline {
  --color-nav-main: var(--tw-color-flickr-blue);
}
```

And in your html

```html
<div class="tw:bg-nav-main">some text</div>
```
